### PR TITLE
Move `UserAgentGenerator` to `_internal`

### LINKED ISSUE #6064: Add `_internal::UserAgentGenerator`
Thinking beyond the release issue today (i.e. for a future change), does it make sense to move such a helper class to internal, if we expect SDK clients to use it?

_Originally posted by @ahsonkhan in https://github.com/Azure/azure-sdk-for-cpp/pull/6063#discussion_r1786908445_
            

              Yes, we should effectively create an alias in the `_internal` namespace, and use it in the new version of eventhubs. But we probably still have to keep this one in `_detail` forever, for the compatibility with previously released eventhubs.

_Originally posted by @antkmsft in https://github.com/Azure/azure-sdk-for-cpp/pull/6063#discussion_r1786910291_

### PR DESCRIPTION
**Pull request by antkmsft**
Possible take on #6064. Fixes #6064.

We would release Core 1.14.1 the same day as EventHubs 1.0.0-beta.10, and they would go to the same vcpkg PR branch and merged into vcpkg together, so ther won;t be a moment when the build in vcpkg CI is broken.
Timing-wise, we only have till November release the latest, and then there is no release in December, so I propose to release it on Thursday October 10th, partially out of band, at least after all the in-band releases on October 8th. It would still make it to the release notes for October.

### CODE REVIEW

**Review by LarryOsterman — APPROVED** [review]

File: sdk/core/azure-core/CHANGELOG.md
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:7]: We typically don't advertise internal only features in the changelog, since we don't want to encourage end customers to use things in that namespace. They are not guaranteed to be stable.

**Comment by ahsonkhan** [comment]
Considering we have the existing patch, there isn't an end customer impact here to motivate an out of band Core release this month. Imo, we should consider it for the regular November release, rather than rush it out.

This will give us some time to discuss and finalize the design of an internal `UserAgentGenerator` class which all upstream SDKs will use. Currently, it isn't used by any upstream SDK that plans to GA (only eventhubs uses this, why don't others?).

I feel hesitant to ship that as GA this week without settling on how we want SDK authors to use it). It isn't urgent.

**Comment by LarryOsterman** [comment]
> This will give us some time to discuss and finalize the design of an internal `UserAgentGenerator` class which all upstream SDKs will use. Currently, it isn't used by any upstream SDK that has GA'd (only eventhubs uses this atm, why don't others?).

Note that the only SDKs which would need to use this are those which are written to protocols other than HTTP. All HTTP based protocols *MUST* use the TelemetryPolicy. 

That effectively means it's use is limited to EventHubs, ServiceBus and (possibly) EventGrid.

File: sdk/core/azure-core/CHANGELOG.md
Comment (antkmsft) [reply sdk/core/azure-core/CHANGELOG.md:7]: It is for the customer to answer the question, "why am I upgrading the Core"? - "It has internal provisions for the packages for telemetry". Plus, should we release this change alone, it has to have some changelog entry, something has clearly changed, we should summarize what it is, and we are doing that, in vague terms, we are not encouraging to use `_internal`, we don't even mention it by name, as `_internal::UserAgentGeenrator`, for example. The contract hasn't changed - do not use `_internal` if you're a customer. I don't know how we can be more vague with these changelog entries, any ideas for better wording?

Comment (ahsonkhan) [reply sdk/core/azure-core/CHANGELOG.md:11]: > Plus, should we release this change alone, it has to have some changelog entry

We have other changes already so no need to make a CL entry for the sake of it (and we'd not ship a change like this alone).

Rather than being vague as a solution, I recommend removing this entry all together. We don't need to let customers know about internal only changes, it adds noise to the changelog that isn't relevant to them.

Comment (antkmsft) [reply sdk/core/azure-core/CHANGELOG.md:11]: Now that there's a "Bugs Fixed" entry, this change for Core is listed in "Other Changes", it will be less visible in web-published changelogs.

File: sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md
Comment (ahsonkhan) [inline sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md:7]: I don't see this a bug fix, nor something that's relevant to the end user. They will not see any behavioral change here, since we are just moving namespaces for an internal implementation.

Comment (antkmsft) [reply sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md:7]: What change do you propose to have listed in the changelog then? There isn't going to be anything else.
And more generally, when we have only the changes that are not immediately visible, what do you think our approach to changelog entries should be? I think this is appropriate.

File: sdk/core/azure-core/CHANGELOG.md
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:3]: I don't think this change warrants shipping a new out-of-band core this month. We should wait till our regular release cadence.
```suggestion
## 1.15.0-beta.1 (Unreleased)
```

**Comment by antkmsft** [comment]
@ahsonkhan, November release should be fine too.
If we are going to have more features to release as Beta first, we will have to cherry-pick the release then (or make two releases in November).

> This will give us some time to discuss and finalize the design of an internal UserAgentGenerator class which all upstream SDKs will use.

This PR is a place to bring any issues then. If there are no PR comments/suggestions by the date of the release, and only the PR approvals, I'd consider it as finalized.

File: sdk/core/azure-core/CHANGELOG.md
Comment (antkmsft) [reply sdk/core/azure-core/CHANGELOG.md:3]: Regular release cadence is fine, this can't be released as beta though.

File: sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md
Comment (ahsonkhan) [reply sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md:7]: For now, no changes (and hold off on shipping out of band).

We should ship updates to our packages that have meaningful impact or benefit to our end customers, or we are gathering beta feedback of features/designs. Both of which naturally warrant changelog entries that are relevant to the user.

If we have anything else planned that would motivate shipping an update to eventhubs, in the future, we could add those entries at that time.

I'd defer to @RickWinter for whether we need to ship such changes, which a changelog update like this is predicated on.

File: sdk/core/azure-core/CHANGELOG.md
Comment (ahsonkhan) [reply sdk/core/azure-core/CHANGELOG.md]: Really, that's unfortunate. How come?
Is it because eventhubs betas are only present in the main registry?

File: cmake-modules/FolderList.cmake
Comment (ahsonkhan) [inline cmake-modules/FolderList.cmake:34]: Just confirming, will other existing GA'd or last shipped beta packages build successfully with the latest `azure-core-cpp` that has `UserAgentGenerator` moved to a different namespace?

I know eventhubs is the only external package outside of core that is explicitly using this class, and hence the only one impacted by the move.
Is the namespace move totally opaque to other packages, that end up using TelemetryPolicy (like storage, keyvault, datatables)?

I think, by reasoning through the change and current callers, it should be fine. But wanted to double check :)

File: sdk/core/azure-core/inc/azure/core/internal/http/user_agent.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/internal/http/user_agent.hpp:26]: For my edification, why do we need upstream callers (client SDK packages like EventHubs) to pass in this value of `cplusplusValue` as a parameter, explicitly?

Why not have the user-agent implementation in core be the only place where we fetch/read the `__cplusplus` macro value, and set it? This makes the caller's job easy. Do we expect per SDK package differences for this value?

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/internal/http/user_agent.hpp:26]: Oh yes, I remember, we have option fields to set them to different values that isn't tied to the language version.

I didn't realize we ended up using private members on the options struct (with friend) to set it, per package:
https://github.com/Azure/azure-sdk-for-cpp/blob/0e48f058d420b205dbff93a84c49a2d23ad52e62/sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp#L77-L93

File: cmake-modules/FolderList.cmake
Comment (antkmsft) [reply cmake-modules/FolderList.cmake:34]: Yes, only the eventhubs is using UserAgentGenerator, and only that needs updating. Other packages don't need core version dependency bump for this.

File: sdk/core/azure-core/inc/azure/core/internal/http/user_agent.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/internal/http/user_agent.hpp:32]: Turns out the `__cplusplus` value is fixed to `199711L` on MSVC/Windows by default, which would skew our results if left as-is. There's a compiler option we could use to address this, which is what I tried here where we build the SDK packages:
https://github.com/Azure/azure-sdk-for-cpp/pull/6148

I don't know how we solve the issue generally, especially if we are planning to keep this approach, and have compiler definitions for sdk, tests, samples.

Ideally, we'd want to have a unit test to detect this issue across platforms.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/internal/http/user_agent.hpp:32]: I updated the code to fall back to use the value of `_MSVC_LANG`, if customers don't use `/Zc:__cplusplus` when compiling their code with MSVC.

**Review by ahsonkhan — APPROVED** [review]
LGTM

**Pull request merged**
